crates/matrix-sdk/src/room/timeline/mod.rs

```diff
@@ -0,0 +1,217 @@
+    ///
+    /// See [`SignalVecExt`](futures_signals::signal_vec::SignalVecExt) for a
+    /// high-level API on top of [`SignalVec`].
+    pub fn signal(&self) -> impl SignalVec<Item = Arc<TimelineItem>> {
+        self.inner.items.signal_vec_cloned()
+    }
+
+    /// Send a message to the room, and add it to the timeline as a local echo.
+    ///
+    /// For simplicity, this method doesn't currently allow custom message
+    /// types.
+    ///
+    /// If the encryption feature is enabled, this method will transparently
+    /// encrypt the room message if the room is encrypted.
+    ///
+    /// # Arguments
+    ///
+    /// * `content` - The content of the message event.
+    ///
+    /// * `txn_id` - A locally-unique ID describing a message transaction with
+    ///   the homeserver. Unless you're doing something special, you can pass in
+    ///   `None` which will create a suitable one for you automatically.
+    ///     * On the sending side, this field is used for re-trying earlier
+    ///       failed transactions. Subsequent messages *must never* re-use an
+    ///       earlier transaction ID.
+    ///     * On the receiving side, the field is used for recognizing our own
+    ///       messages when they arrive down the sync: the server includes the
+    ///       ID in the [`MessageLikeUnsigned`] field `transaction_id` of the
+    ///       corresponding [`SyncMessageLikeEvent`], but only for the *sending*
+    ///       device. Other devices will not see it.
+    ///
+    /// [`MessageLikeUnsigned`]: ruma::events::MessageLikeUnsigned
+    /// [`SyncMessageLikeEvent`]: ruma::events::SyncMessageLikeEvent
+    pub async fn send(
+        &self,
+        content: AnyMessageLikeEventContent,
+        txn_id: Option<&TransactionId>,
+    ) -> Result<()> {
+        let txn_id = txn_id.map_or_else(TransactionId::new, ToOwned::to_owned);
+        handle_local_event(txn_id.clone(), content.clone(), self.room.own_user_id(), &self.inner);
+
+        // If this room isn't actually in joined state, we'll get a server error.
+        // Not ideal, but works for now.
+        let room = Joined { inner: self.room.clone() };
+
+        let response = room.send(content, Some(&txn_id)).await?;
+        add_event_id(&self.inner, &txn_id, response.event_id);
+
+        Ok(())
+    }
+}
+
+/// A single entry in timeline.
+#[derive(Clone, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum TimelineItem {
+    /// An event or aggregation of multiple events.
+    Event(EventTimelineItem),
+    /// An item that doesn't correspond to an event, for example the user's own
+    /// read marker.
+    Virtual(VirtualTimelineItem),
+}
+
+impl TimelineItem {
+    /// Get the inner `EventTimelineItem`, if this is a `TimelineItem::Event`.
+    pub fn as_event(&self) -> Option<&EventTimelineItem> {
+        match self {
+            Self::Event(v) => Some(v),
+            _ => None,
+        }
+    }
+}
+
+fn find_event(
```

Comment:
Hmm, don't we expect this to be problematic? Do we ever remove items from the timeline? I don't think we do. How many items do we want to support? Don't we want to let the caller configure how many items we keep?

We'll likely need an index that maps from the key to the index in any case. Let's at least add a fixme.

Reply:
I'm a little confused, do you mean this is problematic if we keeps those indices around? The intention of this function is just to find the right index to update, it is only meant to be valid for a very short duration.

We could have a separate index so we don't have to do a linear scan, but that would be more complex and I don't think we need it right now (maybe never if we put a reasonable upper bound on the amount of entries that can be in the timeline).

Reply:
No, I meant the growth of the `MutableVec` is unbounded thus the `rfind()` and in turn the `find_event()` function will take an increasing amount of time in the worst case.